File: scripts/manifest.py
#!/usr/bin/env python3
"""Generate and verify the tracked-file release manifest."""

from __future__ import annotations

import argparse
import hashlib
import os
from pathlib import Path
import subprocess
import sys
import tempfile


MANIFEST_NAME = "MANIFEST.sha256"


class ManifestError(RuntimeError):
    pass


def repository_root() -> Path:
    result = subprocess.run(
        ["git", "rev-parse", "--show-toplevel"],
        check=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    return Path(result.stdout.decode("utf-8").strip())


def tracked_paths(root: Path) -> list[str]:
    result = subprocess.run(
        ["git", "ls-files", "--cached", "-z"],
        cwd=root,
        check=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    paths: list[str] = []
    for raw_path in result.stdout.split(b"\0"):
        if not raw_path:
            continue
        try:
            path = raw_path.decode("utf-8")
        except UnicodeDecodeError as exc:
            raise ManifestError("tracked paths must be valid UTF-8") from exc
        if path == MANIFEST_NAME:
            continue
        if "\n" in path or "\r" in path:
            raise ManifestError(f"tracked path cannot contain a newline: {path!r}")
        if path.startswith("/") or path == ".." or path.startswith("../"):
            raise ManifestError(f"tracked path is not repository-relative: {path!r}")
        paths.append(path)
    return sorted(paths, key=lambda value: value.encode("utf-8"))


def sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as stream:
        for chunk in iter(lambda: stream.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def render_manifest(root: Path, paths: list[str]) -> str:
    lines: list[str] = []
    for path in paths:
        absolute_path = root / path
        if not absolute_path.is_file():
            raise ManifestError(f"tracked file is missing from the worktree: {path}")
        lines.append(f"{sha256(absolute_path)}  ./{path}\n")
    return "".join(lines)


def generate(root: Path) -> None:
    paths = tracked_paths(root)
    rendered = render_manifest(root, paths)
    destination = root / MANIFEST_NAME
    temporary_name: str | None = None
    try:
        with tempfile.NamedTemporaryFile(
            mode="w",
            encoding="utf-8",
            newline="\n",
            dir=root,
            prefix=f".{MANIFEST_NAME}.",
            delete=False,
        ) as temporary:
            temporary.write(rendered)
            temporary_name = temporary.name
        os.chmod(temporary_name, 0o644)
        os.replace(temporary_name, destination)
        temporary_name = None
    finally:
        if temporary_name is not None:
            Path(temporary_name).unlink(missing_ok=True)
    print(f"generated {MANIFEST_NAME}: {len(paths)} tracked files")


def parse_manifest(root: Path) -> dict[str, str]:
    manifest_path = root / MANIFEST_NAME
    if not manifest_path.is_file():
        raise ManifestError(f"{MANIFEST_NAME} is missing")
    entries: dict[str, str] = {}
    for line_number, line in enumerate(
        manifest_path.read_text(encoding="utf-8").splitlines(), start=1
    ):
        if len(line) < 69 or line[64:68] != "  ./":
            raise ManifestError(
                f"{MANIFEST_NAME}:{line_number}: invalid checksum line"
            )
        digest = line[:64]
        path = line[68:]
        if any(character not in "0123456789abcdef" for character in digest):
            raise ManifestError(
                f"{MANIFEST_NAME}:{line_number}: invalid SHA-256 digest"
            )
        if not path or "\n" in path or "\r" in path:
            raise ManifestError(
                f"{MANIFEST_NAME}:{line_number}: invalid tracked path"
            )
        if path in entries:
            raise ManifestError(
                f"{MANIFEST_NAME}:{line_number}: duplicate path: {path}"
            )
        entries[path] = digest
    return entries


def verify(root: Path) -> None:
    expected_paths = tracked_paths(root)
    expected = set(expected_paths)
    entries = parse_manifest(root)
    recorded = set(entries)

    problems: list[str] = []
    for path in sorted(expected - recorded):
        problems.append(f"missing manifest entry: {path}")
    for path in sorted(recorded - expected):
        problems.append(f"manifest entry is not tracked: {path}")
    for path in sorted(expected & recorded):
        absolute_path = root / path
        if not absolute_path.is_file():
            problems.append(f"tracked file is missing from the worktree: {path}")
        elif sha256(absolute_path) != entries[path]:
            problems.append(f"checksum mismatch: {path}")

    if problems:
        raise ManifestError("\n".join(problems))
    print(
        f"validated {MANIFEST_NAME}: "
        f"{len(expected_paths)} tracked files and checksums"
    )


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("command", choices=("generate", "check"))
    args = parser.parse_args()
    try:
        root = repository_root()
        if args.command == "generate":
            generate(root)
        else:
            verify(root)
    except (ManifestError, OSError, subprocess.CalledProcessError) as exc:
        print(f"manifest error: {exc}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/manifest.ts
import { execFileSync } from "node:child_process";
import { createHash, randomUUID } from "node:crypto";
import {
  chmodSync,
  existsSync,
  readFileSync,
  renameSync,
  rmSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { isAbsolute, join } from "node:path";

const MANIFEST_NAME = "MANIFEST.sha256";

class ManifestError extends Error {
  override readonly name = "ManifestError";
}

function repositoryRoot(): string {
  return execFileSync("git", ["rev-parse", "--show-toplevel"], {
    encoding: "utf8",
  }).trim();
}

function trackedPaths(root: string): string[] {
  const output = execFileSync("git", ["ls-files", "--cached", "-z"], {
    cwd: root,
  });
  let decoded: string;
  try {
    decoded = new TextDecoder("utf-8", { fatal: true }).decode(output);
  } catch (error) {
    throw new ManifestError("tracked paths must be valid UTF-8", { cause: error });
  }

  const paths: string[] = [];
  for (const path of decoded.split("\0")) {
    if (path.length === 0 || path === MANIFEST_NAME) {
      continue;
    }
    if (path.includes("\n") || path.includes("\r")) {
      throw new ManifestError(`tracked path cannot contain a newline: ${JSON.stringify(path)}`);
    }
    if (isAbsolute(path) || path === ".." || path.startsWith("../")) {
      throw new ManifestError(`tracked path is not repository-relative: ${JSON.stringify(path)}`);
    }
    paths.push(path);
  }
  return paths.sort((left, right) => Buffer.compare(Buffer.from(left), Buffer.from(right)));
}

function sha256(path: string): string {
  return createHash("sha256").update(readFileSync(path)).digest("hex");
}

function renderManifest(root: string, paths: readonly string[]): string {
  return paths
    .map((path) => {
      const absolutePath = join(root, path);
      if (!existsSync(absolutePath) || !statSync(absolutePath).isFile()) {
        throw new ManifestError(`tracked file is missing from the worktree: ${path}`);
      }
      return `${sha256(absolutePath)}  ./${path}\n`;
    })
    .join("");
}

function generate(root: string): void {
  const paths = trackedPaths(root);
  const destination = join(root, MANIFEST_NAME);
  const temporary = join(root, `.${MANIFEST_NAME}.${process.pid}.${randomUUID()}`);
  try {
    writeFileSync(temporary, renderManifest(root, paths), {
      encoding: "utf8",
      flag: "wx",
      mode: 0o600,
    });
    chmodSync(temporary, 0o644);
    renameSync(temporary, destination);
  } finally {
    rmSync(temporary, { force: true });
  }
  console.log(`generated ${MANIFEST_NAME}: ${paths.length} tracked files`);
}

function parseManifest(root: string): Map<string, string> {
  const manifestPath = join(root, MANIFEST_NAME);
  if (!existsSync(manifestPath) || !statSync(manifestPath).isFile()) {
    throw new ManifestError(`${MANIFEST_NAME} is missing`);
  }
  const entries = new Map<string, string>();
  const text = readFileSync(manifestPath, "utf8");
  const lines = text.length === 0 ? [] : text.replace(/\r?\n$/, "").split(/\r?\n/);
  for (const [index, line] of lines.entries()) {
    const lineNumber = index + 1;
    if (line.length < 69 || line.slice(64, 68) !== "  ./") {
      throw new ManifestError(`${MANIFEST_NAME}:${lineNumber}: invalid checksum line`);
    }
    const digest = line.slice(0, 64);
    const path = line.slice(68);
    if (!/^[0-9a-f]{64}$/.test(digest)) {
      throw new ManifestError(`${MANIFEST_NAME}:${lineNumber}: invalid SHA-256 digest`);
    }
    if (path.length === 0 || path.includes("\n") || path.includes("\r")) {
      throw new ManifestError(`${MANIFEST_NAME}:${lineNumber}: invalid tracked path`);
    }
    if (entries.has(path)) {
      throw new ManifestError(`${MANIFEST_NAME}:${lineNumber}: duplicate path: ${path}`);
    }
    entries.set(path, digest);
  }
  return entries;
}

function verify(root: string): void {
  const expectedPaths = trackedPaths(root);
  const expected = new Set(expectedPaths);
  const entries = parseManifest(root);
  const problems: string[] = [];

  for (const path of [...expected].sort()) {
    if (!entries.has(path)) {
      problems.push(`missing manifest entry: ${path}`);
      continue;
    }
    const absolutePath = join(root, path);
    if (!existsSync(absolutePath) || !statSync(absolutePath).isFile()) {
      problems.push(`tracked file is missing from the worktree: ${path}`);
    } else if (sha256(absolutePath) !== entries.get(path)) {
      problems.push(`checksum mismatch: ${path}`);
    }
  }
  for (const path of [...entries.keys()].sort()) {
    if (!expected.has(path)) {
      problems.push(`manifest entry is not tracked: ${path}`);
    }
  }
  if (problems.length > 0) {
    throw new ManifestError(problems.join("\n"));
  }
  console.log(
    `validated ${MANIFEST_NAME}: ${expectedPaths.length} tracked files and checksums`,
  );
}

function main(args: readonly string[]): number {
  if (args.length === 1 && (args[0] === "-h" || args[0] === "--help")) {
    console.log("usage: manifest.ts {generate|check}");
    return 0;
  }
  if (args.length !== 1 || !["generate", "check"].includes(args[0] ?? "")) {
    console.error("usage: manifest.ts {generate|check}");
    return 2;
  }
  try {
    const root = repositoryRoot();
    if (args[0] === "generate") {
      generate(root);
    } else {
      verify(root);
    }
    return 0;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    console.error(`manifest error: ${message}`);
    return 1;
  }
}

process.exitCode = main(process.argv.slice(2));
